[lldb] Pass language options to `ide::isSourceInputComplete`

hamishknight opened the pull request:
Ensure we account for any language options set in the REPL when determining whether a line is considered complete.

I would add a test case but it doesn't seem like there are currently any existing tests for REPL line completeness/indentation, and it doesn't seem like it would be entirely trivial to setup. I have tested locally this resolves the issue.

Resolves #9430

hamishknight commented:
https://github.com/swiftlang/swift/pull/77072

@swift-ci please test

hamishknight commented:
https://github.com/swiftlang/swift/pull/77072

@swift-ci please test

adrian-prantl commented on lldb/source/Plugins/ExpressionParser/Swift/SwiftREPL.cpp:
```diff
@@ -317,9 +317,12 @@ llvm::StringRef SwiftREPL::GetSourceFileBasename() {
       llvm::MemoryBuffer::getMemBuffer(source));
+  swift::LangOptions empty_opts;
+  auto *swift_ast = getSwiftASTContext();
+  auto &lang_opts = swift_ast ? swift_ast->GetLanguageOptions() : empty_opts;
```
I think it would be okay to just early return if there is no swift_ast, but this is fine, too.

adrian-prantl commented on lldb/source/Plugins/ExpressionParser/Swift/SwiftREPL.cpp:
```diff
@@ -572,82 +576,88 @@ void SwiftREPL::CompleteCode(const std::string &current_code,
+  swift::ModuleDecl *repl_module = nullptr;
+  if (m_completion_module_initialized)
+    repl_module = swift_ast->GetModule(completion_module_info, error);
+  if (repl_module == nullptr) {
```
```suggestion
  if (!repl_module) {
```

hamishknight replied:
I should note this is code that was already there, but yeah good catch, thanks!

adrian-prantl commented on lldb/source/Plugins/ExpressionParser/Swift/SwiftREPL.cpp:
```diff
@@ -572,82 +576,88 @@ void SwiftREPL::CompleteCode(const std::string &current_code,
+                                                                        *repl_module, swift::SourceFileKind::Main, bufferID);
+    repl_module->addFile(*repl_source_file);
+    swift::performImportResolution(*repl_source_file);
+    m_completion_module_initialized = true;
```
```suggestion
    m_completion_module_initialized = true;
    return;
```

hamishknight replied:
Actually wait no this branch assigns to `repl_module`, we want to fall through

adrian-prantl commented on lldb/source/Plugins/ExpressionParser/Swift/SwiftREPL.cpp:
```diff
@@ -572,82 +576,88 @@ void SwiftREPL::CompleteCode(const std::string &current_code,
+    swift::performImportResolution(*repl_source_file);
+    m_completion_module_initialized = true;
+  }
+  if (repl_module) {
```
```suggestion
```

adrian-prantl reviewed (approved):
Mostly looks good, thanks! Minor comments inline

adrian-prantl reviewed (changes requested):
Sorry I just realized this is on the wrong branch. Can you do this on 20240723 and cherry-pick to next?

hamishknight committed:
[lldb] Pass language options to `ide::isSourceInputComplete`

Ensure we account for any language options set in the REPL
when determining whether a line is considered complete.

hamishknight commented:
I'll open a separate cherry-picks for rebranch and next, I would like to land this on `main` on the Swift side though

hamishknight commented:
https://github.com/swiftlang/swift/pull/77072

@swift-ci please test

hamishknight commented:
https://github.com/swiftlang/swift/pull/77072

@swift-ci please test

hamishknight commented:
https://github.com/swiftlang/swift/pull/77072

@swift-ci please test Linux

hamishknight commented:
Heh didn't realize how close we were to rebranching, updated base to stable/20240723

hamishknight commented:
https://github.com/swiftlang/swift/pull/77072

@swift-ci please test

hamishknight commented:
https://github.com/swiftlang/swift/pull/77072

@swift-ci please test macOS

hamishknight commented:
https://github.com/swiftlang/swift/pull/77072

@swift-ci please test macOS
